contracts/contracts/vault/OETHVaultCore.sol

```diff
@@ -1,26 +1,148 @@
+import { IERC20 } from "@openzeppelin/contracts/token/ERC20/IERC20.sol";
+import { IStrategy } from "../interfaces/IStrategy.sol";
+
 /**
  * @title OETH VaultCore Contract
  * @author Origin Protocol Inc
  */
 contract OETHVaultCore is VaultCore {
+    using SafeERC20 for IERC20;
+    using StableMath for uint256;
+
     address public immutable weth;
+    uint256 public wethAssetIndex;
 
     constructor(address _weth) {
         weth = _weth;
     }
 
+    /**
+     * @dev Caches WETH's index in `allAssets` variable.
+     *      Reduces gas usage by redeem by caching that.
+     */
+    function cacheWETHAssetIndex() external onlyGovernor {
+        uint256 assetCount = allAssets.length;
+        for (uint256 i = 0; i < assetCount; ++i) {
+            if (allAssets[i] == weth) {
+                wethAssetIndex = i;
+                break;
+            }
+        }
+
+        require(allAssets[wethAssetIndex] == weth, "Invalid WETH Asset Index");
+    }
+
     // @inheritdoc VaultCore
-    function mint(
+    function _mint(
         address _asset,
         uint256 _amount,
         uint256 _minimumOusdAmount
-    ) external virtual override whenNotCapitalPaused nonReentrant {
+    ) internal virtual override {
         require(_asset == weth, "Unsupported asset for minting");
-        _mint(_asset, _amount, _minimumOusdAmount);
+        require(_amount > 0, "Amount must be greater than 0");
+        require(
+            _amount >= _minimumOusdAmount,
+            "Mint amount lower than minimum"
+        );
+
+        emit Mint(msg.sender, _amount);
+
+        // Rebase must happen before any transfers occur.
+        if (!rebasePaused && _amount >= rebaseThreshold) {
+            _rebase();
+        }
+
+        // Mint oTokens
+        oUSD.mint(msg.sender, _amount);
+
+        // Transfer the deposited coins to the vault
+        IERC20(_asset).safeTransferFrom(msg.sender, address(this), _amount);
+
+        // Auto-allocate if necessary
+        if (_amount >= autoAllocateThreshold) {
+            _allocate();
+        }
+    }
+
+    // @inheritdoc VaultCore
+    function _calculateRedeemOutputs(uint256 _amount)
+        internal
+        view
+        virtual
+        override
+        returns (uint256[] memory outputs)
+    {
+        // Overrides `VaultCore._calculateRedeemOutputs` to redeem with only
+        // WETH instead of LST-mix. Doesn't change the function signature
+        // for backward compatibility
+
+        // Calculate redeem fee
+        if (redeemFeeBps > 0) {
+            uint256 redeemFee = _amount.mulTruncateScale(redeemFeeBps, 1e4);
+            _amount = _amount - redeemFee;
+        }
+
+        // Ensure that the WETH index is cached
+        uint256 _wethAssetIndex = wethAssetIndex;
+        require(
+            allAssets[_wethAssetIndex] == weth,
+            "WETH Asset index not cached"
+        );
+
+        outputs = new uint256[](allAssets.length);
```

Comment:
`allAssets.length` is another storage read. This value could be packed into the same slot as `wethAssetIndex` so there is only one slot read for both `wethAssetIndex` and the number of vaults assets.
I don't think it's worth doing this change at this late stage, though.